### ISSUE TITLE
Green swap support

### ISSUE THREAD

**Issue by GValiente** [libretro/gpsp#235]
Thanks for your work on this core, compatibility is much better than I expected!
Most of [Butano](https://github.com/GValiente/butano) games and examples work without issues.

Would it be too hard to add support for the green swap effect?
I know that it's a niche thing and no commercial games used it, but it would be cool to see it on this core too.

Here's a test ROM and a couple of screenshots:
[Test ROM](https://github.com/libretro/gpsp/files/13797830/green_swap.zip)

![green_swap-0](https://github.com/libretro/gpsp/assets/13147012/5a329c4d-6aae-4a53-8076-5cb6808b5155) ![green_swap-1](https://github.com/libretro/gpsp/assets/13147012/6428d599-b437-489a-9702-e1a52fc82c03)

**Comment by andymcca** [comment]
Ah - is that the undocumented mode in DISPCNT?  Not sure how hard that would be to implement - can't speak for David but I might have a look at some point out of curiosity - although ultimately it's probably a bit far down the priority list given there's still issues with commercial games to resolve!  Will keep you updated here tho.

**Comment by davidgfnet** [comment]
Yeah I am aware of this bit, however since no commercial games used it I did not bother :D 
Do homebrew games use this often? I'm curious what's the usecase. Is it to "mimic" some sort of mosaic effect?
We have been fixing many bugs related to new homebrew (for instance Good Boy has uncovered a bunch of bugs) so any bugs and/or fixes are welcome!

**Comment by GValiente** [comment]
I use it quite a bit, usually to mimic a short full screen blur: 
- [Butano Fighter](https://gvaliente.itch.io/butano-fighter) shows it when the MC is hit.
- [Varooom 3D](https://gvaliente.itch.io/varooom-3d) shows it when the car is hit or when a turbo boost is used.
- [GBA-NICCC](https://youtu.be/HtuQa4g90p4?) uses it to show the black and white antialiased mode.

**Comment by GValiente** [comment]
Some pics:
![bf_off](https://github.com/libretro/gpsp/assets/13147012/3e3be02a-0e18-4291-b08c-eac277109e6d) ![bf_on](https://github.com/libretro/gpsp/assets/13147012/4d82125d-acb6-4429-b730-2b74e5bc4b2c)
![v3d_off](https://github.com/libretro/gpsp/assets/13147012/c15f9e6d-ac24-4113-a529-e685d528f5f3) ![v3d_on](https://github.com/libretro/gpsp/assets/13147012/a5c33884-b3e7-48f8-8402-503e31715eaa)
![gba-niccc_off](https://github.com/libretro/gpsp/assets/13147012/7bdf1970-3d16-4fbe-b743-08827d8d78d3) ![gba-niccc_on](https://github.com/libretro/gpsp/assets/13147012/f6328b41-8b86-4f94-8206-0f4efb4a13bf)

**Comment by davidgfnet** [comment]
Cool! Will give it a look! It might require some work since gpsp tries really hard to optimize stuff like palette writes and conversion (for speed) and there's palette calculations all over the place :D 
It's also a case that should be really optimized for it being disabled, since literally 99% of games dont use it.
Thanks!!

**Comment by GValiente** [comment]
Isn't it just a full screen effect after the final image has been composed?

According to [gbatek](https://problemkaputt.de/gbatek.htm#lcdiodisplaycontrol), green swap "appears to be applied to the final picture (ie. after mixing the separate BG and OBJ layers)."

Anyway, thanks a lot for looking at it.

**Comment by davidgfnet** [comment]
Ah right! For some reason I recalled it being like mosaic (with BG/OBJ bits), but it's just a full effect at the end!
Yeah we can just add some routine at the end, won't be very fast tho, but it's OK since nobody uses it.
I have some plan to rewrite some of these parts in gpsp to make it faster. We currently have color correction (which is also a full-screen effect), and we could merge some of these effects so that they don't require a full screen update at the end (and save a lot of cycles, right now video rendering is 50% of the consumed cpu cycles for the emulator).
Thanks!

**Comment by andymcca** [comment]
Hey @GValiente, I just opened draft PRs that implement Green Swap - PR #260 and PR #261.  I used the mgba implementation as a guide.  Looks pretty straightforward but any chance you can build and test it for me?  It looks correct to me when I try the test ROM you gave in the first message but my eyes aren't what they used to be!!!

**Comment by andymcca** [comment]
https://github.com/user-attachments/assets/146a737c-e4f8-4722-a7ef-85e3a90ce4cd

Just me pressing A at random times on your demo to toggle between AA on/off.  Think it seems to be working ok?

**Comment by GValiente** [comment]
There's no fake antialiasing like with mGBA:

![green_swap](https://github.com/user-attachments/assets/bababe50-b964-4c7b-ab59-3deca656af01)

**Comment by andymcca** [comment]
Hmmm - ok I thought it was because I'd used the wrong color modifiers from the mgba implementation.  I switched these to RGB565 but it still looks wrong?

![Screenshot 2024-08-26 103459](https://github.com/user-attachments/assets/db5bcd1a-8f2f-409a-ae05-1cf2659e0987)

**Comment by GValiente** [comment]
The new video looks better, yeah.

Here's a test ROM, if you want to compare it against a real GBA and other emulators: [green_swap.zip](https://github.com/user-attachments/files/16752074/green_swap.zip)

**Comment by andymcca** [comment]
Thanks - yes I used green_swap.gba primarily to compare gpsp to mgba and Nanoboy when troubleshooting intially.  The results are now pretty much identical since my latest commit to the PR.

**Comment by andymcca** [comment]
mgba
![Screenshot 2024-08-31 122356](https://github.com/user-attachments/assets/2ad5911e-0821-42d5-9c5f-e329d43e736f)

**Comment by andymcca** [comment]
gpsp (with this PR)
![Screenshot 2024-08-31 122614](https://github.com/user-attachments/assets/50d516fc-1037-4768-8219-ab1006867ce5)
